src/fmc.rs

```diff
@@ -0,0 +1,278 @@
+//! HAL for Flexible memory controller (FMC)
+
+// From stm32_fmc
+use stm32_fmc::FmcPeripheral;
+use stm32_fmc::{PinsSdram, Sdram, SdramChip, SdramPinSet, SdramTargetBank};
+
+use crate::rcc::{rec, rec::ResetEnable, CoreClocks};
+use crate::stm32;
+use crate::time::Hertz;
+
+use crate::gpio::gpioa::PA7;
+use crate::gpio::gpiob::{PB5, PB6, PB7};
+use crate::gpio::gpioc::{PC0, PC2, PC3, PC4, PC5, PC6, PC7, PC8};
+use crate::gpio::gpiod::{
+    PD0, PD1, PD10, PD11, PD12, PD13, PD14, PD15, PD3, PD4, PD5, PD6, PD7, PD8,
+    PD9,
+};
+use crate::gpio::gpioe::{
+    PE0, PE1, PE10, PE11, PE12, PE13, PE14, PE15, PE2, PE3, PE4, PE5, PE6, PE7,
+    PE8, PE9,
+};
+use crate::gpio::gpiof::{
+    PF0, PF1, PF11, PF12, PF13, PF14, PF15, PF2, PF3, PF4, PF5,
+};
+use crate::gpio::gpiog::{
+    PG0, PG1, PG10, PG12, PG13, PG14, PG15, PG2, PG3, PG4, PG5, PG6, PG7, PG8,
+    PG9,
+};
+use crate::gpio::gpioh::{
+    PH10, PH11, PH12, PH13, PH14, PH15, PH2, PH3, PH5, PH6, PH7, PH8, PH9,
+};
+use crate::gpio::gpioi::{PI0, PI1, PI10, PI2, PI3, PI4, PI5, PI6, PI7, PI9};
+use crate::gpio::{Alternate, AF12, AF9};
+
+/// Storage type for Flexible Memory Controller and its clocks
+pub struct FMC {
+    fmc: stm32::FMC,
+    _prec: rec::Fmc,
+    fmc_ker_ck: Option<Hertz>,
+}
+
+/// Extension trait for FMC controller
+pub trait FmcExt: Sized {
+    fn fmc(self, prec: rec::Fmc, clocks: &CoreClocks) -> FMC;
+
+    /// A new SDRAM memory via the Flexible Memory Controller
+    fn sdram<BANK: SdramPinSet, PINS: PinsSdram<BANK>, CHIP: SdramChip>(
+        self,
+        pins: PINS,
+        chip: CHIP,
+        prec: rec::Fmc,
+        clocks: &CoreClocks,
+    ) -> Sdram<FMC, CHIP> {
+        let fmc = self.fmc(prec, clocks);
+        Sdram::new(fmc, pins, chip)
+    }
+
+    /// A new SDRAM memory via the Flexible Memory Controller
+    fn sdram_unchecked<CHIP: SdramChip, BANK: Into<SdramTargetBank>>(
+        self,
+        bank: BANK,
+        chip: CHIP,
+        prec: rec::Fmc,
+        clocks: &CoreClocks,
+    ) -> Sdram<FMC, CHIP> {
+        let fmc = self.fmc(prec, clocks);
+        Sdram::new_unchecked(fmc, bank, chip)
+    }
+}
+
+impl FmcExt for stm32::FMC {
+    /// New FMC instance
+    fn fmc(self, prec: rec::Fmc, clocks: &CoreClocks) -> FMC {
+        let clk_sel = prec.get_kernel_clk_mux();
+
+        // Calculate kernel clock
+        let fmc_ker_ck = match clk_sel {
+            rec::FmcClkSel::RCC_HCLK3 => Some(clocks.hclk()),
+            rec::FmcClkSel::PLL1_Q => clocks.pll1_q_ck(),
+            rec::FmcClkSel::PLL2_R => clocks.pll2_r_ck(),
+            rec::FmcClkSel::PER => clocks.per_ck(),
+        };
+
+        FMC {
+            fmc: self,
+            _prec: prec,
+            fmc_ker_ck,
+        }
+    }
+}
+
+unsafe impl Sync for FMC {}
```

Comment:
Why do we need to unsafely implement Sync here? Shouldn't the compiler automatically assign sync to the FMC if it's applicable?

I know there are cases where unsafely implementing the Sync is okay (e.g. I've done it for `shared-bus-rtic`), but it could use some documentation?

Reply:
svd2rust implements only implements `Send` for register blocks, and thus FMC will only be automatically marked `Send` but not `Sync`.

Perhaps there's no reason for the `FmcPeripheral` trait to require us to implement `Sync` at all. There's no obvious reason to share it between threads: the use case is to `.init()` it once and that's it.

Reply:
Yeah, I think the right approach may be to remove the `Sync` requirement from the trait in `stm32-fmc`. The downside of this is that we can't then share the FMC peripheral within various RTIC tasks, but I'm not sure if that's an actual downside, since I wouldn't expect that to be a use case.

I see where this is coming from though, so I could understand a desire to leave it in as well (e.g. tasks may want to dynamically reconfigure the FMC chip?)

Reply:
Agreed that removing the requirement for `Sync` is the right approach - [the use case for `Sync` in RTIC](https://rtic.rs/0.5/book/en/by-example/types-send-sync.html) is actually quite small and I don't think applies here. I'll make a PR to remove it